keps/sig-storage/3314-csi-volume-snapshot-delta/README.md

```diff
@@ -0,0 +1,1187 @@
+they will have to extract the appropriate token from their secret.
+
+The `VolumeSnapshotDelta` listener then issues a GRPC call to the
+`GetVolumeSnapshotDelta` service on the storage provider's CSI plugin sidecar:
+
+![CBT Step 3](./img/cbt-step-03.png)
+
+The original `VolumeSnapshotDelta` resource can be retrieved by using the
+identifier (namespace and name) embedded in the URL path, in case there are
+user-provided input parameters which need to be included in the GRPC request.
+
+It is the storage provider's CSI plugin responsibility to call the provider's
+CBT endpoint and manage the in-between authentication and authorisation
+protocols.
+
+The response payloads are then directly returned to the user from the
+`VolumeSnapshotDelta` listener. This synchronous request/response
+mechanism removes the needs to persist the response payloads in etcd.
+
+Pagination parameters from the storage provider will also be included in the
+listener's response to the user. The user will be responsible for coordinating
+subsequent paginated requests.
+
+The user can then use these changed block metadata to retrieve the actual data
+blocks. The design and implementation of the block data retrieval segment of the
+datapath is slated for future KEPs.
+
+### High Availability Mode
+
+In high availability mode where there may be multiple replicas of the
+`VolumeSnapshotDelta` sidecar, an active/passive leader election process will
+be used to elect a single leader instance, while idling other non-leader
+instances.
+
+Non-leader instances will voluntarily fail their readiness probe to remove
+themselves from the `Service`'s request path. Optionally, the `Service` can be
+deployed with its `sessionAffinity` set to `clientIP` to ensure paginated
+requests from the same origin are always directed back to the same leader pod.
+
+### API Specification
+
+The section describes the specification of proposed API. The Go types of the CRD
+are defined as follows:
+
+```go
+// VolumeSnapshotDelta is a specification for a VolumeSnapshotDelta resource
+type VolumeSnapshotDelta struct {
+  metav1.TypeMeta   `json:",inline"`
+
+  // +optional
+  metav1.ObjectMeta `json:"metadata,omitempty"`
+
+  Spec   VolumeSnapshotDeltaSpec   `json:"spec"`
+
+  // +optional
+  Status VolumeSnapshotDeltaStatus `json:"status,omitempty"`
+}
+
+// VolumeSnapshotDeltaSpec is the spec for a VolumeSnapshotDelta resource
+type VolumeSnapshotDeltaSpec struct {
+  BaseVolumeSnapshotName   string `json:"baseVolumeSnapshotName,omitempty"` // name of the base VolumeSnapshot; optional
+  TargetVolumeSnapshotName string `json:"targetVolumeSnapshotName"`        // name of the target VolumeSnapshot; required
+  Mode                     string `json:"mode,omitempty"`         // default to "block"
+}
+
+// VolumeSnapshotDeltaStatus is the status for a VolumeSnapshotDelta resource
+type VolumeSnapshotDeltaStatus struct {
+  Error        string `json:"error,omitempty"`
+  State        string `json:"state"`
```

Comment:
This needs to be an enum with well defined values.

Reply:
Yeah, this entire section can use some re-work. I will update it to proto3, to be more consistent with `csi.proto`.